### PR TITLE
gss-token(1) improvements and fixes.

### PR DESCRIPTION
**Pull request by elric1**
We present a series of improvements and fixes to gss-token(1). We clean up and restructure the code a bit. We re-implement a couple of flags that were forgot when it was imported and changed to use getargs(). We make the code implement a proper GSS loop rather than assuming that there is only a single input and output token. We provide a mechanism to split up tokens to test new functionality that is being developed. And finally, we implement a way to ask GSS to use a particular mech on the initiator side.

### CODE REVIEW

File: lib/gssapi/gss-token.1 [hunk 1]
```diff
@@ -50,26 +54,54 @@ This can be used to load test the KDC.
...
-loop infinitely in read mode.
-This is to support a multiple round trip GSS mechanism.
+loop infinitely in acceptor mode.
+.It Fl m Ar mech
+specifies the GSS mechanism that will be used in initiator mode.
+If
+.Do ? Dc
+is specified, a list of supported mechanisms will be output and
+.Nm
+will exit.
```
Comment (vdukhovni) [inline lib/gssapi/gss-token.1:76]:
```suggestion
If a mechanism name of
.Do ? Dc
is specified, a list of supported mechanisms will be output and
.Nm
will exit.
```

File: lib/gssapi/gss-token.1 [hunk 2]
```diff
@@ -50,26 +54,54 @@ This can be used to load test the KDC.
...
+and will output a token whenever the GSS mechanism provides one.
+Tokens are base64 encoded and terminated by either two successive
+newlines or one newline and EOF.
+The base64 encoding may be broken up by single newlines which will
+be ignored when read.  No extra whitespace will be ignored.
+.Sh EXAMPLES
+To test a simple GSS mechanism which doesn't require a round trip,
+a single
+.Pa /bin/sh
+pipeline will suffice:
```
Comment (vdukhovni) [inline lib/gssapi/gss-token.1:101]: Should there be mention here of a keytab requirement on the acceptor side?

File: lib/gssapi/gss-token.c [hunk 3]
```diff
@@ -164,45 +166,186 @@ gss_mk_err(OM_uint32 maj_stat, OM_uint32 min_stat, const char *preamble)
...
+	int		 first = 1;
+
+	if (nflag)
+		goto bail;
+
+	/*
+	 * According to RFC 2744 page 25, we simply don't output
+	 * zero length output tokens.
+	 */
+	if (out->length == 0)
```
Comment (vdukhovni) [inline lib/gssapi/gss-token.c:224]:
```suggestion
	if (len == 0)
```

File: lib/gssapi/gss-token.c [hunk 4]
```diff
@@ -164,45 +166,186 @@ gss_mk_err(OM_uint32 maj_stat, OM_uint32 min_stat, const char *preamble)
...
+		return -1;
+	}
+	ret = rk_base64_decode(tmp, in->value);
+	if (ret < 0) {
+		free(in->value);
+		if (errno == EOVERFLOW)
+			fprintf(stderr, "Token is too big\n");
+		else
+			fprintf(stderr, "Token encoding is not valid "
+			    "base64\n");
```
Comment (vdukhovni) [inline lib/gssapi/gss-token.c:295]: Should these exit?  Set `in->value` back to NULL? ...

File: lib/gssapi/gss-token.c [hunk 5]
```diff
@@ -529,6 +619,14 @@ main(int argc, char **argv)
 	argc -= optidx;
 	argv += optidx;
 
+	if (mech) {
+		if (mech[0] == '?' && mech[1] == '\0') {
+			print_all_mechs();
+			exit(0);
+		}
+		global_mech = gss_name_to_oid(mech);
```
Comment (vdukhovni) [inline lib/gssapi/gss-token.c:627]: What happens when the lookup fails?

File: lib/gssapi/gss-token.c (hunk 4 above)
Comment (elric1) [reply lib/gssapi/gss-token.c:295]: Thanks for pointing this out. I noticed a few more problems in this code, actually. I'll push updates.

**Review by vdukhovni — APPROVED** [review]
Looks ok.

**Pull request merged**